README.rst
Introduction
============

This package contains customisations on top of the standard 
`Euphorie <http://packages.python.org/Euphorie>`_ application
as used for the OSHA-OIRA website.



docs/HISTORY.txt
Changelog
=========

0.12 (unreleased)
-----------------

- Nothing changed yet.


0.11 (2011-04-12)
-----------------

- #2611 The identification report should also have page numbers in the bottom
        right and the download date in the top right of each page.
- #2885 Parked risks must also be shown affirmatively


0.10 (2011-04-11)
-----------------

- #2560 Added a new schema field on the Sector obj, statistics_level.
- #2699 Headers of the Legal boxes and also the risk headings in the evaluation
        and identification reports must be in lower case for Greek. 
- #2924 OiRA tools with policy risks that have been actioned, should not
        appear in the "Risks that have NOT been evaluated and do NOT have action
        plans", but instead in the top section.
- #2964 Make sure that the logo is visible on the last report page
- #2611 Lots of changes to the final download report 
- #3002 the word "survey" should not be used anymore
- #2989 Final HTML report headers were dodgy in IE7
- #2914 The Hairdressers in Cyprus tool must be shown when viewing the Swedish sector in English
- #2885 String at the bottom of the final report changed.
        Risks that have been identified as not present should be stated affirmatively. 
        Risks must have their priorities indicated (if set)
- #2560 Added admin-edit form and statistics level field on sectors
- #2752 Fixed default color for published surveys
- #2623 Empty legal boxes should not be displayed.

0.9 (2011-03-10)
----------------

- Two bugfixes (for which there aren't any ticket numbers).
  Both are related to the same problem of bullets sometimes being deeper than 4
  levels in the download forms.
  [jcbrand]


0.8 (2011-03-10)
----------------

- Just a version bump. [jcbrand]


0.7 (2011-03-10)
----------------

- #2367 and #2752: Fixed various color picker problems.
- #2750: OiRA client - Change text [jcbrand]
- #2591: Change text on the company form page [jcbrand]
- #2707: OiRA, client - change text above profile questions [jcbrand]


0.6 (2011-03-04)
----------------

- Merged new translation strings and default values to the .po files. [jcbrand]
- Bugfix in touch_surveys.py external-method. [jcbrand]
- #2649: Use portal_properties to store the survey urls. Fallback to English if
  none found. [jcbrand]


0.5 (2011-03-03)
----------------

- Just a version bump. [jcbrand]


0.4 (2011-03-03)
----------------

- #2649: We will now follow the convention that the different SurveyMonkey language
  URLs will be the base url (English version) plus _de, _nl, etc.
  [jcbrand]
- #2681: Remove header capitalization for Greek language. [jcbrand]
- #2555: The footer for the "contents of tool" .rtf document was changed. Also
  removed the "this risk must still be inventorised statement". [jcbrand]
- #2583: Problem in the sessions after updating and republishing [jcbrand]


0.3 (2011-02-23)
----------------

- during the xml import, langauge values might still include trailing and leading spaces. 
  For the frontpage langauge detection, we need to strip them.
  [pilz]
- Bugfix for identification download report generation. [jcbrand]


0.2 (2011-02-23)
----------------

- (Hopefully) Resolves: #1433 #2231 #2293 #2555 #2556 #2621 #2623 #2649
  [jcbrand]


0.1 (2011-01-26)
----------------

* Initial release



osha/oira/utils.py
from Acquisition import aq_inner
from zope.component import getMultiAdapter
from Products.CMFCore.utils import getToolByName
from euphorie.client.sector import IClientSector
from euphorie.client.utils import WebHelpers
from euphorie.content.survey import ISurvey
from euphorie.decorators import reify
from osha.oira.config import lang_dict # Used in templates...

def remove_empty_modules(ls):
    """ Takes a list of modules and risks.

        Removes modules that don't have any risks in them.
        Modules with submodules (with risks) must however be kept.
    """
    while ls and ls[-1].type == 'module':
        ls = ls[:-1]

    for i in range(1, len(ls)):
        if ls[i].type == 'module':
            if ls[i-1].type == 'module' and \
                    ls[i-1].depth >= ls[i].depth:
                ls[i-1] = None
    return ls


def get_unevaluated_nodes(ls):
    """ Takes a list of modules and risks and removes all risks that have *not* 
        been evaluated and actioned.

        Also remove all modules that have lost all their risks in the pocess
    """
    unevaluated = []
    for n in ls:
        if n.type == 'module':
            unevaluated.append(n)

        elif n.type == 'risk':
            # 2924: Policy risks do not get evaluated (but require action plans)
            if not n.action_plans:
                unevaluated.append(n)
            elif n.probability == 0 and n.risk_type != 'policy':
                unevaluated.append(n)
            elif len(n.action_plans):
                # It's possible that there is an action plan object, but
                # it's not yet fully populated
                if n.action_plans[0] == None or \
                        n.action_plans[0].action_plan == None:
                    unevaluated.append(n)

    unevaluated = remove_empty_modules(unevaluated)
    return [u for u in unevaluated if u != None]


def get_evaluated_nodes(ls):
    """ Takes a list of modules and risks and removes all risks that have been
        evaluated and actioned.

        Also remove all modules that have lost all their risks in the pocess
    """
    evaluated = []
    for n in ls:
        if n.type == 'module':
            evaluated.append(n)

        if n.type == 'risk' and (n.probability != 0 or n.risk_type == 'policy') and len(n.action_plans):
                # It's possible that there is an action plan object, but
                # it's not yet fully populated
                plans = [p.action_plan for p in n.action_plans]
                if plans[0] != None:
                    evaluated.append(n)

    evaluated = remove_empty_modules(evaluated)
    return [e for e in evaluated if e != None]


class OSHAWebHelpers(WebHelpers):
    """ Override Euphorie's webhelpers to add some more utility methods.
    """

    def get_language(self):
        """ """
        context = aq_inner(self.context)
        portal_state = getMultiAdapter(
                                (context, self.request), 
                                name=u'plone_portal_state'
                                )
        return portal_state.language()

    def language_dict(self):
        return lang_dict

    def get_sectors_dict(self):
        """ Returns a dictionary with keys being countries (and int. orgs) that
            have sectors inside them and the values being the available survey
            langauges.

            We use ZCatalog directly to bypass permission checks, otherwise we
            get zero surveys returned for anon users.

            See #2556.
        """ 
        context = aq_inner(self.context)
        sectorsfolder = getattr(context, 'sectors')
        if not sectorsfolder:
            return []

        client = getattr(context, 'client')
        if not client:
            return []

        resp = {}
        catalog = getToolByName(context, 'portal_catalog')
        # Only the countries in the client obj should be considered, as the
        # others are not accessible
        for country in client.objectValues():
            langs = {}
            for sector in country.objectValues():
                if not IClientSector.providedBy(sector):
                    continue
                for survey in sector.objectValues():
                    if not ISurvey.providedBy(survey):
                        continue
                    if getattr(survey, "preview", False):
                        continue
                    # XXX: We strip out the country code to keep things simpler.
                    # Otherwise we will have the same language apearing multiple
                    # times on the front page if the surveys are set in different
                    # variants of the same language code.
                    langs[survey.language.split('-')[0].strip()] = 'dummy'
                    
            # surveys might exist without a language set.
            if langs.has_key(''):
                del langs['']
        
            if langs:
                resp[country.id] = langs.keys()

        return resp
